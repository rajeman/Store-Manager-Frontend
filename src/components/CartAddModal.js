import React from 'react';

export default class CartAddModal extends React.Component {
  state = {
    quantity: undefined
  };
  handleQuantityChange = (e) => {
    const quantity = e.target.value
    this.setState(() => {
      return {
        quantity
      };
    });
  }
  render() {
    return (
      <div id="dialog-background-window">
        <div id="dialog-box">
          <div id="content-header">Add Item to cart?</div>
          <div id="cart-item-quantity">
            <label>Quantity:</label>
            <input type="number" id="product-quantity" pattern="^\+?[1-9][\d]*$" required title="Quantity must be a positive integer" onChange={ (e) => {
              this.handleQuantityChange(e);
            }}></input>
          </div>
          <div className="options-button">
            <input id="cancel-button" className="cancel" type="button" value="Cancel" onClick={this.props.handleRemoveModal}></input>
            <input id="cart-button" className="confirm" type="button" value="Confirm" onClick={() => {
              this.props.handleAddItem(this.state.quantity)
            }}></input>
          </div>
        </div>
      </div>
    )
  }
}

