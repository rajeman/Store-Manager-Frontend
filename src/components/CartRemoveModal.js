import React from 'react';

export default (props) => (
  <div id="dialog-background-window">
    <div id="dialog-box">
      <div id="content-header">Are you sure you want to remove product from cart?</div>
      <div className="options-button">
        <input id="delete-button" className="delete" type="button" value="Delete" onClick={
          props.handleRemoveItem}></input>
        {<input id="cancel-button" className="cancel" type="button" value="Cancel" onClick={
          props.handleCancel} ></input>}
      </div>
    </div>
  </div>
)
