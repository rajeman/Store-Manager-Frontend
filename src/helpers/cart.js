
const verifyProductInCart = (products, id) =>
    products.find((product) => product.product_id === id)

const removeProduct = (products, id) =>
    products.filter((product) => product.product_id != id)

const saveToStorage = (email, userItems) => localStorage.setItem(email, JSON.stringify(userItems))

const getUserItems = (email) =>
    localStorage.getItem(email) !== null ? JSON.parse(localStorage.getItem(email)) : {};

export const addProductToCart = (email, product) => {
    const userItems = getUserItems(email);
    const userProducts = userItems.products != undefined ? userItems.products : [];
    const isProductPresent = verifyProductInCart(userProducts, product.product_id);
    if (isProductPresent) {
        return 'DUPLICATE_ITEM';
    }
    userProducts.push(product);
    saveToStorage(email, {
        ...userItems,
        products: userProducts
    });
    return 'ADD_SUCCESS';
}

export const removeCartProduct = (email, id) => {
    const userItems = getUserItems(email);
    const filteredProducts = removeProduct(userItems.products, id);
    saveToStorage(email, {
        ...userItems,
        products: filteredProducts
    });

    return 'REMOVE_SUCCESS';
}

export const removeAllProducts = (email) => {
    const userItems = getUserItems(email);
    saveToStorage(email, {
        ...userItems,
        products: []
    });

    return 'REMOVE_SUCCESS';
}

export const fetchCartProducts = (email) =>
  localStorage.getItem(email) !== null ? JSON.parse(localStorage.getItem(email)).products : [];
