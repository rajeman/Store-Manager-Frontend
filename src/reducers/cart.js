
const defaultState = {
  cartAddModalState: '',
  checkoutState: '',
  checkoutError: ''
};

export default (state = defaultState, action) => {
  switch (action.type) {
    case 'SET_ADD_CART_MODAL':
      return {
        ...state,
        cartAddModalState: action.cartAddModalState
      }
    case 'SET_CHECK_OUT_STATE':
      return {
        ...state,
        checkoutState: action.checkoutState
      }
    case 'SET_CHECK_OUT_ERROR':
      return {
        ...state,
        checkoutError: action.checkoutError
      }
    default:
      return state
  }
};





