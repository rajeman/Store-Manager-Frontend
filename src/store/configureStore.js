import { createStore, combineReducers, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import productReducer from '../reducers/products';
import authReducer from '../reducers/auth';
import saleReducer from '../reducers/records';
import navigationReducer from '../reducers/navigation';
import attendantReducer from '../reducers/attendants';
import cartReducer from '../reducers/cart';

export default () => createStore(
  combineReducers({
    products: productReducer,
    auth: authReducer,
    navigation: navigationReducer,
    sales: saleReducer,
    attendants: attendantReducer,
    cart: cartReducer,
  }), applyMiddleware(thunk),
);
