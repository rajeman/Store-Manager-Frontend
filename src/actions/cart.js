import axios from 'axios';
import { history } from '../routers/AppRouter';
import { getToken } from '../helpers/auth';
import { toast } from 'react-toastify';
import { removeAllProducts } from '../helpers/cart';
import paths from '../helpers/paths';

const url = `${process.env.API_URL}/api/v1/sales`;

export const setAddCartModal = (state) => (
  {
    type: 'SET_ADD_CART_MODAL',
    cartAddModalState: state
  }
)

export const setCheckoutState = (state) => (
  {
    type: 'SET_CHECK_OUT_STATE',
    checkoutState: state
  }
)

export const setCheckoutError = (error) => (
  {
    type: 'SET_CHECK_OUT_ERROR',
    checkoutError: error
  }
)

export const checkoutProducts = (products, email) => dispatch => {
  dispatch(setCheckoutState('STATE_CHECKING_OUT'));
  const headers = {
    headers: { "Authorization": `Bearer ${getToken()}` }
  }
  const reqBody = {
    products
  }

  return axios.post(url, reqBody,
    headers
  )
    .then(({ data }) => {
      dispatch(setCheckoutState('STATE_CHECKOUT_SUCCESS'));
      toast.success('Successfully checked out', {
        hideProgressBar: true
      });
      removeAllProducts(email)
      history.push(`${paths.records}/${data.orderDetails.time_checked_out}`);
    })
    .catch(error => {

      dispatch(setCheckoutState('STATE_CHECKOUT_FAILED'));

      if (!error.response) {
        toast.error('Network Error', {
          hideProgressBar: true
        });
        dispatch(setCheckoutError('Network Error'));
      } else {
        toast.error(error.response.data.error, {
          hideProgressBar: true
        });
        dispatch(setCheckoutError(error.response.data.error));
      }
    });

}
