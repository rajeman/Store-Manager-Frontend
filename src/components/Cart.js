import React from 'react';
import { connect } from 'react-redux';
import CartItem from './CartItem';
import CartRemoveModal from './CartRemoveModal';
import { fetchCartProducts, removeCartProduct } from '../helpers/cart';
import { checkoutProducts } from '../actions/cart';

class Cart extends React.Component {
  state = {
    cartRemoveModal: 'STATE_INVISIBLE',
    selectedProductId: undefined
  }

  handleCheckout = () => {
    const { email } = this.props.auth.userDetails;
    const productsToCheckout = fetchCartProducts(email).map((item) => ({
      productQuantity: item.product_quantity,
      productId: item.product_id
    }));
    this.props.dispatch(checkoutProducts(productsToCheckout, email));
  }

  handleClick = (id) => {
    this.setState(() => {
      return {
        cartRemoveModal: 'STATE_VISIBLE',
        selectedProductId: id
      }
    });
  }

  handleCancelModal = () => {
    this.setState(() => {
      return {
        cartRemoveModal: 'STATE_INVISIBLE'
      }
    });
  }

  handleRemoveItem = () => {
    const remove = removeCartProduct(this.props.auth.userDetails.email,
      this.state.selectedProductId);
    if (remove === 'REMOVE_SUCCESS') {
      this.setState(() => {
        return {
          cartRemoveModal: 'STATE_INVISIBLE'
        }
      });
    }
  };
  render() {
    const { email } = this.props.auth.userDetails;
    const products = fetchCartProducts(email) != undefined ? fetchCartProducts(email) : [];
    return (
      <div className="container">
        <div className="wrapper">
          <div className="container">
            <div className="wrapper">
              {products.length != 0 && <span className="tip">Click on item to remove</span>}
              {products.length != 0 && <input type="button" className="confirm checkout" value="checkout" onClick={this.handleCheckout}></input>}
              <div className="items-box">
                {products.map((product) =>
                  <CartItem item={product} handleClick={this.handleClick} key={product.product_id} />
                )}
              </div>
            </div>
          </div>
        </div>
        {this.state.cartRemoveModal === 'STATE_VISIBLE' && <CartRemoveModal handleCancel={this.handleCancelModal}
          handleRemoveItem={this.handleRemoveItem} />}
      </div>
    );
  }
}

const mapStateToProps = (state) => state;

export default connect(mapStateToProps)(Cart);

