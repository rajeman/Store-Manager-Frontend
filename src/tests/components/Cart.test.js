import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import axios from 'axios';
import Cart from '../../components/Cart';
import CartRemoveModal from '../../components/CartRemoveModal';
import CartAddModal from '../../components/CartAddModal';
import * as CartHelpers from '../../helpers/cart';

jest.mock('axios');
jest.mock('../../helpers/cart');

describe('CART TEST SUITE', () => {
  describe('CART COMPONENT TEST SUITE', () => {
    it('should render cart component', () => {
      const initialState = {
        auth: {
          userDetails: {
            email: 'attendant@attendant.com'
          }
        },
      };
      const mockStore = configureStore([thunk]);
      const store = mockStore(initialState);
      const component = mount(
        <Provider store={store}>
          <Cart />
        </Provider>
      );
      expect(component).toMatchSnapshot();
    });
  });

  it('should checkout cart items if checkout button is clicked', () => {
    const initialState = {
      auth: {
        userDetails: {
          email: 'attendant@attendant.com'
        },
      },
    };
    CartHelpers.fetchCartProducts.mockImplementation(() => [{
      product_id: 12,
      product_name: 'Arduino Mega'
    }]);
    const response = {
      data:
      {
        message: 'successfully modified product'
      }
    };
    axios.post.mockImplementation(() => Promise.resolve(response));
    const mockStore = configureStore([thunk]);
    const store = mockStore(initialState);
    const component = mount(
      <Provider store={store}>
        <Cart />
      </Provider>
    );
    component.find('input[type="button"]').simulate('click', {});
    const dispatchedActions = store.getActions();
    expect(dispatchedActions[0].checkoutState).toEqual('STATE_CHECKING_OUT');
  });

  it('should render the deleteProduct modal', () => {
      const component = mount(<CartRemoveModal />);
      expect(component.exists()).toBe(true);
  });

  it('should render the addProduct modal', () => {
      const component = mount(<CartAddModal />);
      const spy = jest.spyOn(component.instance(), 'handleQuantityChange');
      component.find('input[type="number"]').simulate('change', {});
      expect(spy).toHaveBeenCalled();
  });
});


