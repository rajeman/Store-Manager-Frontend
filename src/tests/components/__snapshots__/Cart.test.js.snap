// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CART TEST SUITE CART COMPONENT TEST SUITE should render cart component 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(Cart)>
    <Cart
      auth={
        Object {
          "userDetails": Object {
            "email": "attendant@attendant.com",
          },
        }
      }
      dispatch={[Function]}
    >
      <div
        className="container"
      >
        <div
          className="wrapper"
        >
          <div
            className="container"
          >
            <div
              className="wrapper"
            >
              <div
                className="items-box"
              />
            </div>
          </div>
        </div>
      </div>
    </Cart>
  </Connect(Cart)>
</Provider>
`;
